File: GradualConstraints/src/de/unifreiburg/cs/proglang/jgs/jimpleutils/Vars.java
package de.unifreiburg.cs.proglang.jgs.jimpleutils;

import de.unifreiburg.cs.proglang.jgs.jimpleutils.Var;
import de.unifreiburg.cs.proglang.jgs.signatures.Symbol;
import soot.Local;
import soot.Value;
import soot.ValueBox;
import soot.jimple.AbstractJimpleValueSwitch;
import soot.jimple.ParameterRef;
import soot.jimple.ThisRef;

import java.util.Collection;
import java.util.stream.Stream;

/**
 * Created by fennell on 3/31/16.
 */
public class Vars {

    /**
     * see getAll(Value)
     */
    public static Stream<Var<?>> getAllFromValues(Collection<Value> bs) {
        return bs.stream().flatMap(Vars::getAll);
    }

    /**
     * see getAll(Value)
     */
    public static Stream<Var<?>> getAllFromValueBoxes(Collection<ValueBox> bs) {
        return bs.stream().flatMap((ValueBox b) -> getAll(b));
    }

    /**
     * see getAll(Value)
     */
    public static Stream<Var<?>> getAll(ValueBox b) {
        return getAll(b.getValue());
    }

    /**
     * Collect all variables from a jimple Value
     */
    public static Stream<Var<?>> getAll(Value val) {
        Stream.Builder<Var<?>> result = Stream.builder();
        val.apply(new AbstractJimpleValueSwitch() {

            @Override
            public void caseThisRef(ThisRef v) {
                result.add(Var.fromThis(v));
            }

            @Override
            public void caseParameterRef(ParameterRef v) {
                result.add(Var.fromParam(Symbol.param(v.getIndex())));
            }

            @Override
            public void caseLocal(Local l) {
                result.add(Var.fromLocal(l));
            }

            @SuppressWarnings("unchecked")
            @Override
            public void defaultCase(Object object) {
                Value v = (Value) object;
                v.getUseBoxes().forEach(o -> {
                    ValueBox b = (ValueBox) o;
                    b.getValue().apply(this);
                });
            }
        });
        return result.build();
    }
}

File: GradualConstraints/src/de/unifreiburg/cs/proglang/jgs/jimpleutils/Vars.scala
package de.unifreiburg.cs.proglang.jgs.jimpleutils

import de.unifreiburg.cs.proglang.jgs.jimpleutils.Var
import de.unifreiburg.cs.proglang.jgs.signatures.Symbol
import soot.Local
import soot.Value
import soot.ValueBox
import soot.jimple.AbstractJimpleValueSwitch
import soot.jimple.ParameterRef
import soot.jimple.ThisRef

import scala.collection.JavaConversions._
import scala.collection.mutable.ListBuffer

/**
  * Created by fennell on 3/31/16.
  */
object Vars {
  /**
    * see getAll(Value)
    */
  def getAllFromValues(bs: java.util.Collection[Value]): Iterator[Var[_]] = {
    return bs.iterator().flatMap(Vars.getAll)
  }

  /**
    * see getAll(Value)
    */
  def getAllFromValueBoxes(bs: java.util.Collection[ValueBox]): Iterator[Var[_]] = {
    return bs.iterator().flatMap((b : ValueBox) => getAll(b))
  }

  /**
    * see getAll(Value)
    */
  def getAll(b: ValueBox): Iterator[Var[_]] = {
    return getAll(b.getValue)
  }

  /**
    * Collect all variables from a jimple Value
    */
  def getAll(value: Value): Iterator[Var[_]] = {
    val result: ListBuffer[Var[_]] = ListBuffer()
    value.apply(new AbstractJimpleValueSwitch() {
      override def caseThisRef(v: ThisRef) {
        result.add(Var.fromThis(v))
      }

      override def caseParameterRef(v: ParameterRef) {
        result.add(Var.fromParam(Symbol.param(v.getIndex)))
      }

      override def caseLocal(l: Local) {
        result.add(Var.fromLocal(l))
      }

      @SuppressWarnings(Array("unchecked")) override def defaultCase(`object`: AnyRef) {
        val v: Value = `object`.asInstanceOf[Value]
        v.getUseBoxes.foreach(o => {
          o.asInstanceOf[ValueBox].getValue().apply(this);
        })
      }
    })
    return result.iterator
  }
}